src/server/transaction.cc

```diff
@@ -844,6 +844,13 @@ void Transaction::ExecuteAsync() {
   // IsArmedInShard in other threads.
   run_count_.store(unique_shard_cnt_, memory_order_release);
 
+  EngineShard* shard = EngineShard::tlocal();
+  if (unique_shard_cnt_ == 1 && shard != nullptr && shard->shard_id() == unique_shard_id_) {
+    DVLOG(1) << "Short-circuit ExecuteAsync " << DebugId();
+    shard->PollExecution("exec_cb", this);
+    return;
+  }
```

Comment:
You should move the use_count_ increment below your fast-path (see lines 833 and 890)

Reply:
This causes `DCHECK(trans != head);` (engine_shard_set.cc:386) to fail in many tests.

Reply:
hmm,  @dranikpg has a point. I am afraid you have a leak when you increase use_count_ like this,.

Reply:
but now I understand the complexity of this change. I think that run_count_ should also move. we need to deep dive into the logic there and it can not be done during these hours :)

Reply:
Yes, we should carefully look into it because the use_count_ mechanism is a little fragile. But I'm more than sure that just decrementing it (instead of moving the increment below) will work 😄

Reply:
Shouldn't I be calling `intrusive_ptr_release(this);` instead of just decrementing, to make sure that it's deleted?
BTW I don't understand how this mechanism works with multi/lua, how are transactions not destroyed after every single operation? We don't use `run_count_.fetch_add()` but `.store()` instead..

Reply:
there is `run_count_` and there is `use_count_`.  #jobsecurity

Reply:
Ok, I added an explicit call to `intrusive_ptr_release(this);`.
@romange @dranikpg - please be extra pedantic :)